code/modules/photography/camera/camera.dm

```diff
@@ -0,0 +1,221 @@
+		pictures_left = pictures_max
+		return
+	if(istype(I, /obj/item/disk/holodisk))
+		if (!disk)
+			if(!user.transferItemToLoc(I, src))
+				to_chat(user, "<span class='warning'>[I] is stuck to your hand!</span>")
+				return TRUE
+			to_chat(user, "<span class='notice'>You slide [I] into the back of [src].</span>")
+			disk = I
+		else
+			to_chat(user, "<span class='warning'>There's already a disk inside [src].</span>")
+		return TRUE //no afterattack
+	..()
+
+/obj/item/camera/examine(mob/user)
+	..()
+	to_chat(user, "It has [pictures_left] photos left.")
+
+/obj/item/camera/proc/can_target(atom/target, mob/user, prox_flag)
+	if(!on || blending || !pictures_left || (!isturf(target) && !isturf(target.loc)) || !((isAI(user) && GLOB.cameranet.checkTurfVis(get_turf(target))) || ((user.client && (get_turf(target) in get_hear(user.client.view, user)) || (get_turf(target) in get_hear(world.view, user))))))
+		return FALSE
+	return TRUE
+
+/obj/item/camera/afterattack(atom/target, mob/user, flag)
+	if (disk)
+		if(ismob(target))
+			if (disk.record)
+				QDEL_NULL(disk.record)
+
+			disk.record = new
+			var/mob/M = target
+			disk.record.caller_name = M.name
+			disk.record.set_caller_image(M)
+		else
+			to_chat(user, "<span class='warning'>Invalid holodisk target.</span>")
+			return
+
+	if(!can_target(target, user, flag))
+		return
+
+	on = FALSE
+	addtimer(CALLBACK(src, .proc/cooldown), cooldown)
+	icon_state = state_off
+
+	INVOKE_ASYNC(src, .proc/captureimage, target, user, flag, picture_size_x, picture_size_y)
+
+
+/obj/item/camera/proc/cooldown()
+	UNTIL(!blending)
+	icon_state = state_on
+	on = TRUE
+
+/obj/item/camera/proc/show_picture(mob/user, datum/picture/selection)
+	var/obj/item/photo/P = new(src, selection)
+	P.show(user)
+	to_chat(user, P.desc)
+	qdel(P)
+
+/obj/item/camera/proc/captureimage(atom/target, mob/user, flag, size_x = 1, size_y = 1)
+	blending = TRUE
+	var/turf/target_turf = get_turf(target)
+	if(!isturf(target_turf))
+		blending = FALSE
+		return FALSE
+	size_x = CLAMP(size_x, 0, CAMERA_PICTURE_SIZE_HARD_LIMIT)
+	size_y = CLAMP(size_y, 0, CAMERA_PICTURE_SIZE_HARD_LIMIT)
+	var/list/desc = list()
+	var/ai_user = isAI(user)
+	var/list/seen
+	var/list/viewlist = (user && user.client)? getviewsize(user.client.view) : getviewsize(world.view)
+	var/viewr = max(viewlist[1], viewlist[2]) + max(size_x, size_y)
+	var/viewc = user.client? user.client.eye : target
+	seen = get_hear(viewr, viewc)
+	var/list/turfs = list()
+	var/list/mobs = list()
+	var/blueprints = FALSE
+	var/clone_area = SSmapping.RequestBlockReservation(size_x * 2 + 1, size_y * 2 + 1)
+	for(var/turf/T in block(locate(target_turf.x - size_x, target_turf.y - size_y, target_turf.z), locate(target_turf.x + size_x, target_turf.y + size_y, target_turf.z)))
+		if((ai_user && GLOB.cameranet.checkTurfVis(T)) || T in seen)
+			turfs += T
+			for(var/mob/M in T)
+				mobs += M
+			if(locate(/obj/item/areaeditor/blueprints) in T)
+				blueprints = TRUE
+			CHECK_TICK

```

Comment:
The whole point of the reservation thing was so this CHECK_TICK wouldn't be necessary